src/main/kotlin/org/partiql/scribe/sql/RexConverter.kt

```diff
@@ -159,78 +164,258 @@ public open class RexConverter(
+                    noPrecision -> error("Invalid PType in plan $pType has a scale but no precision specified")
+                    else -> DataType.NUMERIC(pType.precision, pType.scale)
+                }
+            }
+            PType.DECIMAL -> {
+                val noPrecision = pType.unspecifiedPrecision()
+                val noScale = pType.unspecifiedScale()
+                when {
+                    noPrecision && noScale -> DataType.DECIMAL()
+                    noScale -> DataType.DECIMAL(pType.precision)
+                    noPrecision -> error("Invalid PType in plan $pType has a scale but no precision specified")
+                    else -> DataType.DECIMAL(pType.precision, pType.scale)
+                }
+            }
+            // Approximate numeric types
+            PType.REAL -> DataType.REAL()
+            PType.DOUBLE -> DataType.DOUBLE_PRECISION()
+            // String types
+            PType.CHAR ->
+                when (pType.unspecifiedLength()) {
+                    true -> DataType.CHAR()
+                    false -> DataType.CHAR(pType.length)
+                }
+            PType.VARCHAR ->
+                when (pType.unspecifiedLength()) {
+                    true -> DataType.VARCHAR()
+                    false -> DataType.VARCHAR(pType.length)
+                }
+            PType.STRING -> DataType.STRING()
+            // Datetime types
+            PType.DATE -> DataType.DATE()
+            PType.TIME ->
+                when (pType.unspecifiedPrecision()) {
+                    true -> DataType.TIME()
+                    false -> DataType.TIME(pType.precision)
+                }
+            PType.TIMEZ ->
+                when (pType.unspecifiedPrecision()) {
+                    true -> DataType.TIME_WITH_TIME_ZONE()
+                    false -> DataType.TIME_WITH_TIME_ZONE(pType.precision)
+                }
+            PType.TIMESTAMP ->
+                when (pType.unspecifiedPrecision()) {
+                    true -> DataType.TIMESTAMP()
+                    false -> DataType.TIMESTAMP(pType.precision)
                 }
-                PType.DECIMAL -> {
-                    val noPrecision = rex.target.unspecifiedPrecision()
-                    val noScale = rex.target.unspecifiedScale()
-                    when {
-                        noPrecision && noScale -> DataType.DECIMAL()
-                        noScale -> DataType.DECIMAL(rex.target.precision)
-                        noPrecision -> error("Invalid PType in plan ${rex.target} has a scale but no precision specified")
-                        else -> DataType.DECIMAL(rex.target.precision, rex.target.scale)
+            PType.TIMESTAMPZ ->
+                when (pType.unspecifiedPrecision()) {
+                    true -> DataType.TIMESTAMP_WITH_TIME_ZONE()
+                    false -> DataType.TIMESTAMP_WITH_TIME_ZONE(pType.precision)
+                }
```

Comment:
self-review: this chunk is moved from `visitCast`. The remaining part w/ intervals is new